### DISCUSSION TITLE
number of subsets versus permutations

### DISCUSSION THREAD

**Discussion by markusloecher** [shap/shap#1498] [Q&A] [answered]
I am struggling with a conceptual issue which has not become very clear to me from the papers that I have read:
Shapley values average over **all possible feature subsets** (which there are $2^M$ of) whereas SHAP values seem to average over **all possible feature orderings**, which there are $M!$ of. This distinction seems important, especially since the entire discussion on “TreeExplainer” (SHAP values for trees) has a strong focus on the ordering of the nodes. 
Any insight on bridging these two combinatorial views would be greatly appreciated.

Thanks,
Markus

**Comment by slundberg** [comment, marked as answer]
As you point out there are two ways to write the Shapley value formula:

1) By averaging over all subsets:

<img src="https://user-images.githubusercontent.com/3740613/95140961-d9400280-0724-11eb-8a24-88386e7b0b86.png" width=400 />

and 2) by averaging over all permutations: 

<img src="https://user-images.githubusercontent.com/3740613/95141066-14423600-0725-11eb-9444-8fd5a48f6d88.png" width=340 />

(where you can [check Wikipedia](https://en.wikipedia.org/wiki/Shapley_value) for the meaning of all the symbols)

Notice that while each formula is an average, the permutation based approach is a standard unweighted average, where each term has equal weight, while the set-based formulation is a weighted average that essentially counts how many permutations would produce each set.

Different explainers inside SHAP use each of those different formulations. The set based one has (as you point out) "only" $2^M$ sets to evaluate (where M is the number of input features), which is less than the $M!$ orderings. So `shap.explainers.Exact` uses the set based formulation and completely enumerates all the sets (Kernel explainer also uses the set one). In contrast `shap.explainers.Permutation` uses the permutation formulation because it enables a fast sampling procedure based on toggling one feature at a time while always guaranteeing "efficiency" or local accuracy (meaning `shap_values.sum(1) + shap_values.base_value == model(X)`).

Since `shap.explainers.Tree` is an exact explainer specific to trees, it doesn't really matter which formula you use (since we are not sampling from it), it is really just about which formulation makes the algorithm easier to understand. We chose to use the permutation version for the algorithm description given in our [Nature MI paper](https://rdcu.be/b0z70).

**Comment by markusloecher** [comment]
Thanks a lot, this is very helpful. I did not realize that the original formulation of the Shapley values already had a permutation version.

**Reply by tevfikaytekin** [reply]
By the way, these two formulations are equivalent.
